packages/core/src/extension/extension-messenger.ts

```diff
@@ -0,0 +1,101 @@
+import { IframeEventPoster } from '../common/iframe';
+import { appendWww, parseUrl } from '../common/url';
+import { bindDecorator as bind } from '../common/utility';
+
+import { ExtensionNotFoundError } from './errors';
+import { Extension } from './extension';
+import { ExtensionCommandHandlers } from './extension-command-handler';
+import { ExtensionOriginEvent } from './extension-origin-event';
+import { HostOriginEvent } from './host-origin-event';
+
+export class ExtensionMessenger {
+    private _isListening = false;
+    private _extensions: Extension[] | undefined;
+    private _extensionCommandHandlers: ExtensionCommandHandlers = {};
+    private _posters: { [extensionId: string]: IframeEventPoster<HostOriginEvent> } = {};
+
+    listen(extensions: Extension[], extensionCommandHandlers: ExtensionCommandHandlers): void {
+        this._extensions = extensions;
+        this._extensionCommandHandlers = extensionCommandHandlers;
+
+        if (this._isListening) {
+            return;
+        }
+
+        this._isListening = true;
+
+        window.addEventListener('message', this._handleMessage);
```

Comment:
Is there a particular reason why you chose to listen to messages directly rather than using `IframeEventListener`?

Reply:
I thought it might be too costly to have a separate `IframeEventListener` and command handlers for each extension at beginning. Now I have replaced it with the existing `IframeEventListener`.